server/src/main/java/org/elasticsearch/node/NodeConstruction.java

```diff
@@ -969,7 +969,7 @@ record PluginServiceInstances(
         nodeService = new NodeService(
             settings,
             threadPool,
-            new MonitorService(settings, nodeEnvironment, threadPool),
+            new MonitorService(settings, clusterService.getClusterSettings(), nodeEnvironment, threadPool),
```

Comment:
This uses the `DiskThresholdSettings` according to the local node, but that's not necessarily correct, it's the settings on the elected master node which matter. They'll be the same if the settings are set via `PUT _cluster/settings`, but different if set in `elasticsearch.yml`.

Reply:
If the settings are not there in `_cluster/settings` then they will come from the `elasticsearch.yml` _on the elected master_. Here you're looking at the settings from the `elasticsearch.yml` _on the local node_, and these things may be different.

Reply:
Hi @DaveCTurner , sorry I might misunderstanding the key point.
>then they will come from the elasticsearch.yml on the elected master.

It should not come from elected master's yml? It should come from the yml of the local node itself?
>Here you're looking at the settings from the elasticsearch.yml on the local node

The above `DiskThresholdSettings` is construncted based on `settings` and `clusterSettings` in above FsService.java L47:
`final FsProbe probe = new FsProbe(nodeEnvironment, new DiskThresholdSettings(settings, clusterSettings));`
So it could evaluate static configure and dynamic cluster settings.

I have done a real cluster integration test:
1. Setup a two nodes cluster, both have data and master role.
2. No dynamic cluster settings, and set one node's watermark as `70%, 75%, 80%`, another node keep default. Get node stats API result, they are different:
```
 "nodes" : {
    "oKcA8bgMSCeUK1vb0cOyfA" : {
      "fs" : {
        "timestamp" : 1712733138988,
        "total" : {
          "total_in_bytes" : 21003583488,
          "free_in_bytes" : 18922135552,
          "available_in_bytes" : 18261114880
        },
        "data" : [
          {
            "path" : "/data1/elasticsearch-8.13.0/data",
            "mount" : "/data1 (/dev/vdb)",
            "type" : "ext4",
            "total_in_bytes" : 21003583488,
            "free_in_bytes" : 18922135552,
            "available_in_bytes" : 18261114880,
            "effective_low_watermark" : "70%",
            "effective_high_watermark" : "75%",
            "effective_flood_stage" : "80%"
          }
        ],
...
   "data" : [
          {
            "path" : "/data1/elasticsearch-8.13.0/data",
            "mount" : "/data1 (/dev/vdb)",
            "type" : "ext4",
            "total_in_bytes" : 21003583488,
            "free_in_bytes" : 18706587648,
            "available_in_bytes" : 18045566976,
            "effective_low_watermark" : "85%",
            "effective_high_watermark" : "90%",
            "effective_flood_stage" : "95%"
          }

```
3. Update cluster settings:
```
PUT localhost:9200/_cluster/settings
{
  "persistent": {
    "cluster.routing.allocation.disk.watermark.low": "50%",
    "cluster.routing.allocation.disk.watermark.high": "60%",
    "cluster.routing.allocation.disk.watermark.flood_stage": "70%"
  }
}'
```
4. Get node stats again, they become the same:
```
        "data" : [
          {
            "path" : "/data1/elasticsearch-8.13.0/data",
            "mount" : "/data1 (/dev/vdb)",
            "type" : "ext4",
            "total_in_bytes" : 21003583488,
            "free_in_bytes" : 18922033152,
            "available_in_bytes" : 18261012480,
            "effective_low_watermark" : "50%",
            "effective_high_watermark" : "60%",
            "effective_flood_stage" : "70%"
          }
        ],
...
        "data" : [
          {
            "path" : "/data1/elasticsearch-8.13.0/data",
            "mount" : "/data1 (/dev/vdb)",
            "type" : "ext4",
            "total_in_bytes" : 21003583488,
            "free_in_bytes" : 18706501632,
            "available_in_bytes" : 18045480960,
            "effective_low_watermark" : "50%",
            "effective_high_watermark" : "60%",
            "effective_flood_stage" : "70%"
          }
        ],
```

So this should be expected? Or could you please guide me in the right direction?

Reply:
The issue is that all the allocation decisions are made on the elected master, using the elected master's watermark configuration, so we must ignore the values from `elasticsearch.yml` on all other nodes. Your example shows that we're not ignoring those values here.

Reply:
Got it, so we should always get threshold settings from `clusterSettings` right?

Reply:
No, `clusterSettings` will fall back to the local node's `elasticsearch.yml` if the watermarks are not set in the cluster state which isn't right. We need to reach out to the elected master here. I think I'd extend `org.elasticsearch.action.admin.cluster.allocation.TransportGetAllocationStatsAction` to include this information in its response, since we already invoke this action as part of the node-stats API, but in principle we could add another action just for this info too.

Reply:
Okay, I am going to extend `TransportGetAllocationStatsAction`.

Reply:
Hi @DaveCTurner, I have checked https://github.com/elastic/elasticsearch/pull/105894, so it's better to add threshold watermarks into `NodeAllocationStats`? If user requires `allocations` metric in `_node/stats` then we return threshold watermark stats in this allocations blob?

Reply:
> it's better to add threshold watermarks into `NodeAllocationStats`? 

The exact watermarks depend on both the watermark config (from the elected master) and the actual disk sizes (from the individual nodes). I think it'd be better to include just the watermark config in `org.elasticsearch.action.admin.cluster.allocation.TransportGetAllocationStatsAction.Response` and then have the node coordinating the node stats request combine this information with the disk sizes of the individual nodes according to each node's `FsInfo`.

> If user requires `allocations` metric in `_node/stats` then we return threshold watermark stats in this allocations blob?

I think I preferred your original idea of including this in each node's `FsInfo` section, which is returned if the user requests the `fs` metric.